[UI] Add editor for json parameters in new run page

drewbutlerbb4 opened the pull request:
Fixes #2714 by adding a button that opens an editor if a parameter is recognized as a json object.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/pipelines/2747)
<!-- Reviewable:end -->

Bobgy commented:
Thank you! This is awesome!
I will do a review soon

drewbutlerbb4 commented:
This PR should probably be adapted based on the end result of #2752. I will wait to see what is decided and then update accordingly

Bobgy commented:
@drewbutlerbb4 thanks! https://github.com/kubeflow/pipelines/pull/2752 is approved, you may reference it.

May I ask for a short video demonstrating the new UX, that makes review easier.
One plugin you can try is: https://chrome.google.com/webstore/detail/screencastify-screen-vide/mmeijimgabbpbgpdklnllpncmdofkcpn

Bobgy commented on frontend/src/components/NewRunParameters.tsx:
```diff
@@ -16,18 +16,48 @@
     super(props);
+
+    this.state = {
+      isBeingEdited: {},
```
An early suggestion is: instead of keeping a dictionary of state in `NewRunParameters` component, state will be easier to manage if you define a sub component named maybe `EnhancedTextField`, this component keeps a boolean state of whether editor mode is turned on and the extra logic of switching between an editor and a text field.

The extra layer of abstraction would make both `NewRunParameters` component and the `EnhancedTextField` component easier to read.

Since the new component is only used here, it's fine to keep it in the same file and there's no need to make it very general.

Bobgy replied:
After the change, you may find there are too many snapshot changes... that's a problem of our unit tests (they test too much implementation details). We are moving towards testing components more by `mount` instead of `shallow`.

drewbutlerbb4 replied:
I have reformatted it in the requested way and updated the snapshots

drewbutlerbb4 commented:
[Here](https://drive.google.com/file/d/1zcoB7jgnDt9sTfylHmqhAxerAsTum_TF/view?usp=sharing) is a video demonstration.

drewbutlerbb4 commented:
I have made the requested changes and just need to finish changing the test 'fires handleParamChange callback on change' in NewRunParameters.test.tsx which fails I believe because of the shallow rendering not rendering the children EnhancedTextField components. Any suggestions on how to resolve this would be appreciated.

Bobgy commented:
The video UX looks awesome!

Bobgy commented:
/ok-to-test

Bobgy commented:
> I have made the requested changes and just need to finish changing the test 'fires handleParamChange callback on change' in NewRunParameters.test.tsx which fails I believe because of the shallow rendering not rendering the children EnhancedTextField components. Any suggestions on how to resolve this would be appreciated.

Recommend using `mount` instead of `shallow` when testing behavior (you probably need to update some existing tests). Trigger the change event on the input instead.

Current unit tests are unfriendly with making extra abstractions in the component, that's our existing problem to fix. If you feel there's any difficulties, feel free to reach out to me. You can submit with just the implementation (comment out tests) and I can follow up with unit tests.

Bobgy commented on frontend/src/components/NewRunParameters.tsx:
```diff
@@ -39,23 +61,150 @@ class NewRunParameters extends React.Component<NewRunParametersProps> {
+
+class EnhancedTextField extends React.Component<EnhancedTextFieldProps, EnhancedTextFieldState> {
+
+  public static getDerivedStateFromProps(nextProps: EnhancedTextFieldProps, prevState: EnhancedTextFieldState): EnhancedTextFieldState {
```
Refer to https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops

I don't see a strong reason to use `getDerivedStateFromProps`, you can just put these in render function. That can probably make it easier to understand.

drewbutlerbb4 replied:
I assumed that once an editor option has been added to the Textfield then that editor option would always remain regardless of if the current input was a json string or not. My reasoning for using `getDerivedStateFromProps` is that we need to update the state if this is the first time a json string has been identified, but doing so in the render would cause the render to be impure. Do you think there is a better lifecycle method for this or do you still want me to put it in the render function?

Bobgy replied:
The behavior is well thought over. I agree with it too.

I overlooked getDerivedState using previous state. Thanks for the explanation, this should be a suitable use case.

Bobgy commented on frontend/src/components/NewRunParameters.tsx:
```diff
@@ -39,23 +61,150 @@ class NewRunParameters extends React.Component<NewRunParametersProps> {
+                        isEditorOpen: !this.state.isEditorOpen,
+                        isInJsonForm: this.state.isInJsonForm,
+                        isJsonField: this.state.isJsonField
+                      });
```
setState receives a patch and updates state based on it, there's no need to specify isInJsonForm and isJsonField if they don't need to change.

drewbutlerbb4 commented:
I have updated the failing test, added a new test, and fixed some formatting issues. Hopefully the new test is acceptable.

Thanks for the comments and reviews!

Bobgy commented on frontend/src/components/NewRunParameters.tsx:
```diff
@@ -39,23 +61,145 @@ class NewRunParameters extends React.Component<NewRunParametersProps> {
+            variant='outlined'
+            label={param.name}
+            value={param.value || ''}
+            onChange={ev => handleParamChange(index, ev.target.value || '')}
```
nit: these calls to `handleParamChange` with an index is leaking implementation details. EnhancedTextField should be a complete component that can be used by itself (although it's fine we put them in the same file for now).

Recommend renaming the prop to `onChange` and accepts just the value.

Bobgy commented on frontend/src/components/NewRunParameters.tsx:
```diff
@@ -39,23 +61,145 @@ class NewRunParameters extends React.Component<NewRunParametersProps> {
+      return (
+        <>
+          <TextField
+            id={`newRunPipelineParam${index}`}
```
nit: same as below reasoning, it would probably make EnhancedTextField more encapsulated if it also accepts an id prop that will be passed on here.

Bobgy commented on frontend/src/components/NewRunParameters.tsx:
```diff
@@ -39,23 +61,145 @@ class NewRunParameters extends React.Component<NewRunParametersProps> {
+                    }}
+                    style={{ padding: '3px 5px', margin: 0 }}
+                  >
+                    {this.state.isEditorOpen ? 'Close Editor' : 'Open Editor'}
```
I'm not sure on this one. To be accurate, this is a JSON editor, not a general one. Do you think it's worth clarifying that?

drewbutlerbb4 replied:
I think clarifying would help the user understand why the button is popping up, however it looks a little clunky. For this I think you are right and the value of an easy to understand design outweighs the cost of it looking somewhat clunky. I'll add it for now and can remove it if you change your mind

Bobgy replied:
Sounds good, I still prefer clarity:
1. it's a new feature
2. the button is hidden initially, only when parameter is JSON shows it. After the button is shown, opening the editor would be the best way to edit JSON. so it doesn't matter how much space the button takes.

Bobgy commented on frontend/src/components/NewRunParameters.tsx:
```diff
@@ -39,23 +61,145 @@ class NewRunParameters extends React.Component<NewRunParametersProps> {
 }
 
+interface EnhancedTextFieldProps {
+  param: ApiParameter;
```
Because you are naming props as param and handleParamChange, shall we rename the component to ParamEditor?

Bobgy commented on frontend/src/pages/NewRun.tsx:
```diff
@@ -798,6 +798,13 @@ class NewRun extends Page<{}, NewRunState> {
           p.value = (p.value || '').trim();
+          // Remove any special formatting for displaying jsons on an editor
+          try {
+            const parsedJson = JSON.stringify(JSON.parse(p.value || ''));
```
Is this more like `formattedJson`? I'm curious why would this be useful. Can you also add a test case for this in NewRun.test.tsx?

drewbutlerbb4 replied:
Yes, I think `unformattedJson` would be best. If you leave the json editor open and click start run then p.value will include all of the newline and whitespace characters used to display the json in the editor. This wouldn't be an issue for most standardized json parsers, but could be an issue for a custom parser. 

Another option would be to add `formattedJson` in the state for ParamEditor which holds the json string with any formatting done in `getDerivedStateFromProps` which would remove the need for any checks in NewRun.tsx. But this would mean an extra parse and stringify would be needed onChange and there would be a some redundancy between props and state. I prefer the original solution, but I will defer to you.

Bobgy replied:
I don't agree on this. Here's my reasoning:

* With the auto formatting at pipeline submit time, a user doesn't know the parameter is formatted before it is sent out. If there's something unexpected happens that don't work with formatted result. It would be really hard to identify the root cause.
* Without auto formatting, a user knows the parameter is exactly the same as in his editor on the UI. If something wrong happens, the user will know they could either close the editor to restore it to one line. Or fix their own system that doesn't handle standard JSON.

In general, I think KFP UI shouldn't secretly help user preprocess parameter. What the user sees on the UI should be the value we send to APIs.

drewbutlerbb4 replied:
Okay that makes sense, I removed it from NewRun.tsx

Bobgy commented on frontend/src/components/NewRunParameters.test.tsx:
```diff
@@ -37,6 +37,19 @@ describe('NewRunParameters', () => {
+      titleMessage: 'Specify json parameters required by the pipeline',
+    } as NewRunParametersProps;
+    const tree = mount(<NewRunParameters {...props} />);
+    tree.find('Button#chooseExperimentBtn').simulate('click');
```
reminder, did you forget to push your commit?

drewbutlerbb4 replied:
It's pushed now

Bobgy commented on frontend/src/components/NewRunParameters.test.tsx:
```diff
@@ -37,6 +37,20 @@ describe('NewRunParameters', () => {
+      titleMessage: 'Specify json parameters required by the pipeline',
+    } as NewRunParametersProps;
+    const tree = mount(<NewRunParameters {...props} />);
+    tree.find('Button#toggleEditorBtn').simulate('click');
```
Hey, sorry I missed a problem.
id should be unique in the web page, but there can be multiple editor buttons.

This id is only for testing purpose, a widely accepted convention is to use attribute `data-testid` as an identifier if a better selector cannot be found.

Bobgy replied:
Recommend finding the button by its text:
https://github.com/kubeflow/pipelines/blob/e824c0b2f63a8891a18f437a1ad2b94990dd67a9/frontend/src/components/viewers/Tensorboard.test.tsx#L147
the hostNodes command filters to only html nodes.

(that's also how users find a button)

Bobgy commented on frontend/src/components/NewRunParameters.tsx:
```diff
@@ -39,23 +65,143 @@ class NewRunParameters extends React.Component<NewRunParametersProps> {
+      isInJsonForm: false,
+      isJsonField: false,
+    };
+  }
```
suggestion: no need to use a constructor
```
class AAA {
  state = {
    isEditorOpen: false,
    isInJsonForm: false,
    isJsonField: false,
  }
}
```

Bobgy commented on frontend/src/components/NewRunParameters.tsx:
```diff
@@ -39,23 +65,143 @@ class NewRunParameters extends React.Component<NewRunParametersProps> {
+            label={param.name}
+            value={param.value || ''}
+            onChange={ev => onChange(ev.target.value || '')}
+            style={{ maxWidth: 600 }}
```
nit: Please write this as a css
you can follow the pattern https://github.com/kubeflow/pipelines/blob/e824c0b2f63a8891a18f437a1ad2b94990dd67a9/frontend/src/components/SideNav.tsx#L262 to concatenate two class names

Bobgy commented on frontend/src/components/NewRunParameters.test.tsx:
```diff
@@ -37,6 +37,20 @@ describe('NewRunParameters', () => {
     expect(shallow(<NewRunParameters {...props} />)).toMatchSnapshot();
   });
 
+  it('clicking the open editor button for json parameters display an editor', () => {
```
May I ask for one more test case covering also typing in the json Editor and submit the run?

drewbutlerbb4 commented:
cc @animeshsingh
